src/mem_table.rs

```diff
@@ -172,46 +175,38 @@ impl KVTable {
         MemTableIterator(self.map.range(bounds))
     }
 
-    /// Puts a value, returning as soon as the value is written to the memtable but before
+    /// Inserts a value, returning as soon as the value is written to the memtable but before
     /// it is flushed to durable storage.
-    fn put(&self, key: Bytes, value: Bytes, attrs: RowAttributes) {
-        self.maybe_subtract_old_val_from_size(key.clone());
+    fn put_or_delete(&self, key: Bytes, value: ValueDeletable, attrs: RowAttributes) {
+        let key_len = key.len();
+        let value_len = match &value {
+            ValueDeletable::Tombstone => 0,
+            ValueDeletable::Value(v) => v.len(),
+        };
         self.size.fetch_add(
-            key.len() + value.len() + sizeof_attributes(&attrs),
+            key_len + value_len + sizeof_attributes(&attrs),
             Ordering::Relaxed,
         );
-        self.map.insert(
-            key,
-            ValueWithAttributes {
-                value: ValueDeletable::Value(value),
-                attrs,
-            },
-        );
-    }
-
-    fn delete(&self, key: Bytes, attrs: RowAttributes) {
-        self.maybe_subtract_old_val_from_size(key.clone());
-        self.size
-            .fetch_add(key.len() + sizeof_attributes(&attrs), Ordering::Relaxed);
-        self.map.insert(
-            key,
-            ValueWithAttributes {
-                value: ValueDeletable::Tombstone,
-                attrs,
-            },
-        );
-    }
-
-    fn maybe_subtract_old_val_from_size(&self, key: Bytes) {
-        if let Some(old_deletable) = self.get(&key) {
-            let old_size = key.len()
-                + match old_deletable.value {
-                    ValueDeletable::Tombstone => 0,
-                    ValueDeletable::Value(old) => old.len(),
-                }
-                + sizeof_attributes(&old_deletable.attrs);
-            self.size.fetch_sub(old_size, Ordering::Relaxed);
-        }
+        let once = OnceCell::new();
+        self.map
+            .compare_insert(key, ValueWithAttributes { value, attrs }, |old_value| {
+                // Compare fn might be called multiple times, so we use `OnceCell`
+                // to avoid calling `subtract_old_val_from_size` multiple times.
+                once.get_or_init(|| {
+                    self.subtract_old_val_from_size(key_len, old_value);
+                });
```

Comment:
I'm not convinced that OnceCell solves our problem. Here's the race I'm worried about:

0. key1 -> val0
1. Thread1: compare_insert(key1, val1)
2. Thread2: compare_insert(key1, val2)
3. Thread1: replace with old_value=val0 -> success (subtracts val0.size, first get)
4. Thread2: replace with old_value=val0 -> failure (subtracts val0.size, first get)
5. Thread2: replace with old_value=val1 -> success (does nothing, second get)

We want (5) to succeed the second time, not the first time.

Reply:
Yeah that sounds right to me. I think the guarantee we can rely on is that the last time `compare_and_insert` is called you'll have the value that it's replacing. You can use that to write the old value to some local holder (I think a `RefCell` should be fine) and then update the size after `compare_and_insert` returns.

Reply:
fixed

Reply:
All three of your points are good! Didn't even think about the fact that the current impl has the same problem.